src/reactors/UnifiedReactor.sol

```diff
@@ -0,0 +1,181 @@
+    using CurrencyLibrary for address;
+
+    /// @notice thrown when an auction resolver is not set
+    error EmptyAuctionResolver();
+    /// @notice thrown when an order's nonce has already been used
+    error InvalidNonce();
+    /// @notice thrown when the order targets a different reactor
+    error InvalidReactor();
+    /// @notice thrown when the order's deadline has passed
+    error DeadlinePassed();
+
+    /// @notice Permit2 instance for signature verification and token transfers
+    IPermit2 public immutable permit2;
+
+    constructor(IPermit2 _permit2, address _protocolFeeOwner) ProtocolFeesV2(_protocolFeeOwner) {
+        permit2 = _permit2;
+    }
+
+    /// @inheritdoc IReactor
+    function execute(SignedOrder calldata order) external payable override nonReentrant {
+        ResolvedOrderV2[] memory resolvedOrders = new ResolvedOrderV2[](1);
+        resolvedOrders[0] = _resolve(order);
+        _prepare(resolvedOrders);
+        _fill(resolvedOrders);
+    }
+
+    /// @inheritdoc IReactor
+    function executeBatch(SignedOrder[] calldata orders) external payable override nonReentrant {
+        uint256 ordersLength = orders.length;
+        ResolvedOrderV2[] memory resolvedOrders = new ResolvedOrderV2[](ordersLength);
+
+        unchecked {
+            for (uint256 i = 0; i < ordersLength; i++) {
+                resolvedOrders[i] = _resolve(orders[i]);
+            }
+        }
+
+        _prepare(resolvedOrders);
+        _fill(resolvedOrders);
+    }
+
+    /// @inheritdoc IReactor
+    function executeWithCallback(SignedOrder calldata order, bytes calldata callbackData)
+        external
+        payable
+        override
+        nonReentrant
+    {
+        ResolvedOrderV2[] memory resolvedOrders = new ResolvedOrderV2[](1);
+        resolvedOrders[0] = _resolve(order);
+        _prepare(resolvedOrders);
+        IReactorCallbackV2(msg.sender).reactorCallback(resolvedOrders, callbackData);
+        _fill(resolvedOrders);
+    }
+
+    /// @inheritdoc IReactor
+    function executeBatchWithCallback(SignedOrder[] calldata orders, bytes calldata callbackData)
+        external
+        payable
+        override
+        nonReentrant
+    {
+        uint256 ordersLength = orders.length;
+        ResolvedOrderV2[] memory resolvedOrders = new ResolvedOrderV2[](ordersLength);
+
+        unchecked {
+            for (uint256 i = 0; i < ordersLength; i++) {
+                resolvedOrders[i] = _resolve(orders[i]);
+            }
+        }
+
+        _prepare(resolvedOrders);
+        IReactorCallbackV2(msg.sender).reactorCallback(resolvedOrders, callbackData);
+        _fill(resolvedOrders);
+    }
+
+    /// @notice Resolve a SignedOrder into a ResolvedOrderV2 using the auction resolver
+    function _resolve(SignedOrder calldata signedOrder) internal view returns (ResolvedOrderV2 memory resolvedOrder) {
+        (address auctionResolver, bytes memory orderData) = abi.decode(signedOrder.order, (address, bytes));
```

Comment:
alternatively, can add the auctionResolver field to the SignedOrder struct:
```
struct SignedOrder {
    bytes order;
    address auctionResolver;
    bytes sig;
}
```
}